### PR TITLE
[Components] Implemented Deferrable Server

### PR DESCRIPTION
**Pull request by NicolaGiardino**
## 拉取/合并请求描述：(PR description)

[This PR as been submitted because of my needing for a deferrable server. I implemented one for this RTOS, and it has been tested intensively both with the QEMU emulator and the Raspberry Pi Pico]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in the submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use a web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [x] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality
- [x] 本拉取/合并使用[formatting](https://github.com/mysterywolf/formatting)等源码格式化工具确保格式符合[RT-Thread代码规范](../documentation/coding_style_cn.md) This PR complies with [RT-Thread code specification](../documentation/coding_style_en.txt)

### CODE REVIEW

**Comment by mysterywolf** [comment]
Thank you for your PR, we still need a few days to review this PR.😀
BTW, would you like to help us to translate the [readme of homepage](https://github.com/RT-Thread/rt-thread/blob/master/README.md) to your native language? 
image

**Comment by NicolaGiardino** [comment]
Thank you for your answer, I'll wait for your feedback.
I'll start drafting something out, but I can't guarantee anything about timings due to other commitments

**Comment by mysterywolf** [comment]
> Thank you for your answer, I'll wait for your feedback. I'll start drafting something out, but I can't guarantee anything about timings due to other commitments

No problem, thank you~

File: src/deferrable.c [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+ * This function will stop the deferrable server when its capacity is finished
+ *
+ * @param NULL parameter passed to the timer function
+ *
+ * @return void
+ */
+void rt_timer_deferrable_capacity(void *param)
+{
+    deferrable_server.capacity_left = 0;
+    rt_thread_suspend(&deferrable_server.deferrable);
```
Comment (mysterywolf) [inline src/deferrable.c:34]: Do you want suspend other thread or suspend the thread itself?
rt-thread's suspend API only can be invoked in a thread and suspend this thread itself. You cannot invoke rt_thread_suspend API

Comment (NicolaGiardino) [reply src/deferrable.c:34]: This is invoked only while the server is the current running thread, so it suspends itself

**Comment by mysterywolf** [comment]
related paper
[1995_-_J_K_Strosnider_-_Thedeferrableserveralgorithmforenhancedaperiodicre[retrieved_2021-12-23].pdf](https://github.com/RT-Thread/rt-thread/files/7767831/1995_-_J_K_Strosnider_-_Thedeferrableserveralgorithmforenhancedaperiodicre.retrieved_2021-12-23.pdf)
:

**Comment by NicolaGiardino** [comment]
> related paper:
> [1995_-_J_K_Strosnider_-_Thedeferrableserveralgorithmforenhancedaperiodicre[retrieved_2021-12-23].pdf](https://github.com/RT-Thread/rt-thread/files/7767831/1995_-_J_K_Strosnider_-_Thedeferrableserveralgorithmforenhancedaperiodicre.retrieved_2021-12-23.pdf)
> 

Sorry, is this for knowledge or is there something wrong in the code? 😅

**Comment by mysterywolf** [comment]
> > related paper:
> > [1995_-_J_K_Strosnider_-_Thedeferrableserveralgorithmforenhancedaperiodicre[retrieved_2021-12-23].pdf](https://github.com/RT-Thread/rt-thread/files/7767831/1995_-_J_K_Strosnider_-_Thedeferrableserveralgorithmforenhancedaperiodicre.retrieved_2021-12-23.pdf)
> 
> Sorry, is this for knowledge or is there something wrong in the code? 😅

This just gives a reference. We still need a few days to assess this code. You made a big improvement

File: include/deferrable.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+*/
+
+#include <rtthread.h>
+
+#define DEBUG_DEFERRABLE 0
+
+#define DEFERRABLE_SERVER_STACK_SIZE 4096
+
+rt_uint8_t rt_deferrable_stack[DEFERRABLE_SERVER_STACK_SIZE];
+
```
Comment (NicolaGiardino) [inline include/deferrable.h:26]: Sorry for the lack of comments in the .h file, if you need any clarifications just ask..

Comment (mysterywolf) [reply include/deferrable.h:26]: Don't worry. You don't need to fix that. leave that to me. I will modify them. I am on it.

Comment (mysterywolf) [reply include/deferrable.h:26]: RT-Thread has a [workqueue](https://github.com/RT-Thread/rt-thread/blob/master/components/drivers/src/workqueue.c). Please consider what different between the workqueue and yours.

File: src/deferrable.c (hunk 1 above)
Comment (mysterywolf) [reply src/deferrable.c:34]: Nop, it cannot be invoked by current thread. Timers will run in another thread or directly run in interruptes.

Comment (NicolaGiardino) [reply src/deferrable.c:34]: Okay, I understand, how can this be done instead? Is there a way to do something similar from another thread

**Comment by mysterywolf** [comment]
The Deferrable server has a higher performance of real-time than workqueue, but because of the `rt_thread_suspend` issue, I has labeled this PR for "the future version". Thank you for you contribution ahead.😁

**Comment by NicolaGiardino** [comment]
> The Deferrable server has a higher performance of real-time than workqueue, but because of the `rt_thread_suspend` issue, I has labeled this PR for "the future version". Thank you for you contribution ahead.😁

Thank you, I'm getting on it and update this commit with a version without the susped call asap

**Comment by mysterywolf** [comment]
I have helped you to fix some format issues. Please consider to merge my commit: https://github.com/mysterywolf/rt-thread/tree/deferr.

**Comment by NicolaGiardino** [comment]
I can't find the files in your fork

**Comment by mysterywolf** [comment]
Thank you for your contribution, but we cannot merge this. We will improve workqueue and combine the method of deferrable server in it in future.

**Pull request closed without merging**